runtime/src/main/java/io/quarkiverse/messaginghub/pooled/jms/PooledJmsDecorator.java

```diff
@@ -32,46 +35,72 @@ private ConnectionFactory getConnectionFactory() {
+        return factory;
     }
 
     @Override
     public Connection createConnection() throws JMSException {
-        return getConnectionFactory().createConnection();
+        return isInJmsPoolConnectionFactory() ? delegate.createConnection() : getConnectionFactory().createConnection();
     }
 
     @Override
     public Connection createConnection(final String userName, final String password) throws JMSException {
-        return getConnectionFactory().createConnection(userName, password);
+        return isInJmsPoolConnectionFactory() ? delegate.createConnection(userName, password)
+                : getConnectionFactory().createConnection(userName, password);
     }
 
     @Override
     public JMSContext createContext() {
-        return runInContext(ConnectionFactory::createContext);
+        return getConnectionFactory().createContext();
     }
 
     @Override
     public JMSContext createContext(final String userName, final String password) {
-        return runInContext(cf -> cf.createContext(userName, password));
+        return getConnectionFactory().createContext(userName, password);
     }
 
     @Override
     public JMSContext createContext(final String userName, final String password, final int sessionMode) {
-        return runInContext(cf -> cf.createContext(userName, password, sessionMode));
+        return getConnectionFactory().createContext(userName, password, sessionMode);
     }
 
     @Override
     public JMSContext createContext(final int sessionMode) {
-        return runInContext(cf -> cf.createContext(sessionMode));
+        return getConnectionFactory().createContext(sessionMode);
+    }
+
+    @Override
+    public XAConnection createXAConnection() throws JMSException {
+        return isInJmsPoolConnectionFactory() ? delegate.createXAConnection()
+                : ((XAConnectionFactory) getConnectionFactory()).createXAConnection();
+    }
+
+    @Override
+    public XAConnection createXAConnection(final String userName, final String password) throws JMSException {
+        return isInJmsPoolConnectionFactory() ? delegate.createXAConnection(userName, password)
+                : ((XAConnectionFactory) getConnectionFactory()).createXAConnection(userName, password);
+    }
+
+    @Override
+    public XAJMSContext createXAContext() {
+        return ((XAConnectionFactory) factory).createXAContext();
+    }
+
+    @Override
+    public XAJMSContext createXAContext(final String userName, final String password) {
+        return ((XAConnectionFactory) factory).createXAContext(userName, password);
     }
 
-    private JMSContext runInContext(Function<ConnectionFactory, JMSContext> process) {
-        ConnectionFactory cf = getConnectionFactory();
-        isContext = true;
-        try {
-            return process.apply(cf);
-        } finally {
-            isContext = false;
+    private boolean isInJmsPoolConnectionFactory() {
+        for (StackTraceElement stack : Thread.currentThread().getStackTrace()) {
```

Comment:
The error loks like
```
Caused by: java.lang.StackOverflowError
		at io.quarkiverse.messaginghub.pooled.jms.PooledJmsDecorator.createXAConnection(PooledJmsDecorator.java:104)
		at io.quarkiverse.messaginghub.pooled.jms.it.CustomConnectionFactory_Subclass.createXAConnection(Unknown Source)
		at org.apache.activemq.artemis.jms.client.ActiveMQConnectionFactory.createXAConnection(ActiveMQConnectionFactory.java:359)
		at io.quarkiverse.messaginghub.pooled.jms.it.CustomConnectionFactory_Subclass.createXAConnection$$superforward(Unknown Source)
		at io.quarkiverse.messaginghub.pooled.jms.it.PooledJmsDecorator_3082110ee669b1ff5deeed85a3a77d97b4e27fd5_Delegate_Subclass.createXAConnection(Unknown Source)
		at org.messaginghub.pooled.jms.JmsPoolXAConnectionFactory.createProviderConnection(JmsPoolXAConnectionFactory.java:85)
		at org.messaginghub.pooled.jms.JmsPoolXAConnectionFactory.createXAConnection(JmsPoolXAConnectionFactory.java:178)

```

Reply:
I think that the problem here is that `ActiveMQConnectionFactory#createXAConnection()` in turn calls `ActiveMQConnectionFactory#createXAConnection(final String username, final String password)` which is also intercepted/decorated (at least in ArC; the CDI spec is not really clear when it comes to interception of self-invocation).

You may try to use the `StackWalker` API instead of `Thread#getStackTrace()` :shrug:.

Reply:
@mkouba @osmman I refator to use StackWalker API and can you take a review again? Thanks!